### ISSUE TITLE
read_namespaced_pod_log with follow=True hangs when there is a long gap between logs from pod

### ISSUE THREAD

**Issue by 0xdholleran** [kubernetes-client/python#1234]
**What happened (please include outputs or screenshots)**:
If I try to read the logs of a pod with the following:
```
from kubernetes import config, client
config.load_incluster_config()
k8s = client.CoreV1Api()
namespace = "default"
pod_name="watch-this"
logs = k8s.read_namespaced_pod_log(
    name=pod_name,
    namespace=namespace,
    follow=True,
    _preload_content=False,
)
for line in logs:
  print(line)
```

 and there is a gap (say 5 minutes)  between line1 and line2 of the log being written, then the function hangs after the first line

**What you expected to happen**:

the function should read all the logs as they are written from the watched pod and then exit once finished

**How to reproduce it (as minimally and precisely as possible)**:

Create the following pod to watch:
```
apiVersion: v1
kind: Pod
metadata:
  name: watch-this
  namespace: default
spec:
  containers:
  - command:
    - python3
    - -u
    - -c
    - |
      from time import sleep
      for i in range(3):
        print(i)
        if i != 2:
          sleep(300)
    image: python:3.7.4-slim-stretch
    imagePullPolicy: IfNotPresent
    name: python
    resources: {}
    stdin: true
    stdinOnce: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
  priority: 0
  restartPolicy: Never
  schedulerName: default-scheduler
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
```

then watch the pod with the following:
```
kubectl -n default run -i -t test-python-kube-client --image=python:3.7.4-slim-stretch --restart=Never --command -- /bin/sh

pip install kubernetes==11.0.0
python
from kubernetes import config, client
config.load_incluster_config()
k8s = client.CoreV1Api()
namespace = "default"
pod_name="watch-this"
logs = k8s.read_namespaced_pod_log(
    name=pod_name,
    namespace=namespace,
    follow=True,
    _preload_content=False,
)
for line in logs:
  print(line)
```

the output should get stuck like this:
`b'0\n'`

however if I run the equivalent request using kubectl proxy and the kubernetes API, it works as expected:
```
kubectl proxy --port=8080 &
curl http://localhost:8080/api/v1/namespaces/default/pods/watch-this/log?follow=true
0
1
2
```
**Anything else we need to know?**:

**Environment**:
- Kubernetes version (`Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:11:31Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.13", GitCommit:"1da71a35d52fa82847fd61c3db20c4f95d283977", GitTreeState:"clean", BuildDate:"2020-07-15T21:59:26Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}`):
- OS ( MacOS 10.14.6):
- Python version (`3.7.4`)
- Python client version (`kubernetes==11.0.0`)

**Comment by roycaihw** [comment]
/assign
cc @palnabarun

**Comment by roycaihw** [comment]
You can use the watch util to stream logs. For example: https://github.com/kubernetes-client/python-base/blob/fd322f70aa6c33782ab466afd4eaae271a27cbcb/watch/watch_test.py#L83-L85

**Comment by dholleran-lendico** [comment]
hi @roycaihw, that doesn't work either. If I try to stream the logs of the pod defined above with:
```
from kubernetes import config, client, watch
config.load_incluster_config()
k8s = client.CoreV1Api()
namespace = "default"
pod_name="watch-this"
w = watch.Watch()
for e in w.stream(k8s.read_namespaced_pod_log, name=pod_name, namespace=namespace):
    print(e)
```

This doesn't work either. It also hangs after the first line

**Comment by raphael-bos** [comment]
I have faced the same problem this last week. After spending a fair amount of time investigating, I came to the conclusion the problem is with the TCP Keep Alive socket option, which urllib3 doesn't set by default. So, while running a long http request, the server doesn't get any tcp keep alive probes from the client and decides to close the connection without notifying the client, which then hangs waiting for the complete response from the server.
These socket options worked for me (the times chosen are only for demonstration):
`socket_options= [
    (socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1),
    (socket.IPPROTO_TCP, socket.TCP_KEEPIDLE, 5),
    (socket.IPPROTO_TCP, socket.TCP_KEEPINTVL, 30),
    (socket.IPPROTO_TCP, socket.TCP_KEEPCNT, 3)])`
Socket options should be set in the urllib3.PoolManager class here: https://github.com/kubernetes-client/python/blob/master/kubernetes/client/rest.py#L106

Another side effect of running long tcp connections can be seen in #1158 . Basically, connections used by the PoolManager are never closed and therefore and unusable for subsequently requests. This also makes the client hangs if it tries to make another request (read_namespaced_pod, for example) after a long running one. A dirty hack is to call `rest_client.pool_manager.clear()` after a long running request.
